### PR TITLE
  Extend docker provider api surface to use management components

### PR DESCRIPTION
**Pull request by cxbrowne1207**
*Issue #, if available:*

*Description of changes:*
This PR extends the docker provider api surface to use management component to gradually introduce the changes outlined in [this PR](https://github.com/aws/eks-anywhere/pull/7353), split out into multiple. This is the first one.

*Testing (if applicable):*

*Documentation added/planned (if applicable):*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pkg/cluster/fetch.go [hunk 1]
```diff
@@ -11,6 +11,30 @@ import (
 	v1alpha1release "github.com/aws/eks-anywhere/release/api/v1alpha1"
 )
 
+// ManagementComponents contain the versions bundle for a managment cluster's core components.
+type ManagementComponents struct {
+	*v1alpha1release.VersionsBundle
+}
```
Comment (g-gaston) [inline pkg/cluster/fetch.go:17]: I would just copy over whatever are management components and have fields for it
This doesn't decouple upgraders from the format of a versions bundle

I mean something like

```suggestion
type ManagementComponents struct {
    Version string
	CoreCapi releasev1.Whatever
	DockerProvider releasev1.WhateverProvider
	...
	EKSAComponents releasev1.WhateverManifest
}
```

File: pkg/providers/docker/docker.go [hunk 2]
```diff
@@ -77,7 +77,8 @@ type ProviderKubectlClient interface {
 	UpdateAnnotation(ctx context.Context, resourceType, objectName string, annotations map[string]string, opts ...executables.KubectlOpt) error
 }
 
-func NewProvider(providerConfig *v1alpha1.DockerDatacenterConfig, docker ProviderClient, providerKubectlClient ProviderKubectlClient, now types.NowFunc) providers.Provider {
+// NewProvider returns a new providers.Provider.
+func NewProvider(providerConfig *v1alpha1.DockerDatacenterConfig, docker ProviderClient, providerKubectlClient ProviderKubectlClient, now types.NowFunc) providers.TemporaryProvider {
```
Comment (g-gaston) [inline pkg/providers/docker/docker.go:81]: does this need to return an interface?
can't we just return the actual type?

File: pkg/providers/provider.go [hunk 3]
```diff
@@ -46,6 +46,15 @@ type Provider interface {
 	PreCoreComponentsUpgrade(ctx context.Context, cluster *types.Cluster, clusterSpec *cluster.Spec) error
 }
 
+// TemporaryProvider is a Provider but expanded to take on different interfaces for a few methods. This is a temporary change
+// to facilitate refactoring efforts in smaller changes.
+type TemporaryProvider interface {
```
Comment (g-gaston) [inline pkg/providers/provider.go:51]: do we need a new interface? i think if we return the types, we don't
also, we can still return the old interface even if we don't return the type

what is this interface for? no one is consuming it

Comment (cxbrowne1207) [reply pkg/providers/provider.go:51]: This was added to add the new methods without changing the original interface first. They would not be accessible in the tests through `provider.Provider`. Perhaps, we can export the docker `provider` struct that implements `provider.Provider` and reference that in the tests instead?

File: pkg/providers/docker/docker.go (hunk 2 above)
Comment (cxbrowne1207) [reply pkg/providers/docker/docker.go:81]: We can make it return a type

File: pkg/cluster/fetch.go [hunk 4]
```diff
@@ -11,6 +11,58 @@ import (
 	v1alpha1release "github.com/aws/eks-anywhere/release/api/v1alpha1"
 )
 
+// ManagementComponents contains the versions bundle for a management cluster's core components.
```
Comment (g-gaston) [inline pkg/cluster/fetch.go:14]:
```suggestion
// ManagementComponents bundles the resource definitions of all EKS-A management components.
```

maybe?

File: pkg/providers/docker/docker.go [hunk 5]
```diff
@@ -77,8 +79,9 @@ type ProviderKubectlClient interface {
 	UpdateAnnotation(ctx context.Context, resourceType, objectName string, annotations map[string]string, opts ...executables.KubectlOpt) error
 }
 
-func NewProvider(providerConfig *v1alpha1.DockerDatacenterConfig, docker ProviderClient, providerKubectlClient ProviderKubectlClient, now types.NowFunc) providers.Provider {
-	return &provider{
+// NewProvider returns a new providers.Provider.
```
[the remark this answers is not in the document]
Comment (cxbrowne1207) [reply pkg/providers/docker/docker.go:82]: Oh, I forgot to change this one after I update it to return the concrete type. Will change

File: pkg/cluster/fetch.go (hunk 4 above)
Comment (cxbrowne1207) [reply pkg/cluster/fetch.go:14]: Sounds better. Thanks!

**Pull request merged**